src/librustc_resolve/diagnostics.rs

```diff
@@ -397,6 +397,102 @@ impl Bar {
 ```
 "##,
 
+E0412: r##"
+An undeclared type name was used. Example of erroneous codes:
+
+```
+impl Something {} // error: use of undeclared type name `Something`
+// or:
+trait Foo {
+    fn bar(N); // error: use of undeclared type name `N`
+}
+```
+
+To fix this error, please verify you didn't misspell the type name,
+you did declare it or imported it into the scope. Examples:
+
+```
+struct Something;
+
+impl Something {}
+// or:
+trait Foo {
+    type N;
+
+    fn bar(Self::N);
+}
+```
+"##,
+
+E0413: r##"
+A declaration shadows an enum variant or unit-like struct in scope.
+Example of erroneous code:
+
+```
+struct Foo;
+
+let Foo = 12i32; // error: declaration of `Foo` shadows an enum variant or
+                 //        unit-like struct in scope
+```
+
+
+To fix this error, make sure declarations do not shadow any enum variants or
+structures in the scope. Please also verify that neither name was misspelled.
```

Comment:
I don't think talking about misspelling is appropriate here - after all, the problem is an accidental match. Maybe "rename the variable such that it doesn't shadow any enum variable or structure in scope".